source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc

```diff
@@ -0,0 +1,89 @@
+
+#include <cstdint>
+
+#include "common/common/assert.h"
+
+#include "extensions/io_socket/user_space_io_socket/peer_buffer.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace IoSocket {
+namespace UserSpaceIoSocket {
+
+UserSpaceFileEventImpl::UserSpaceFileEventImpl(Event::Dispatcher& dispatcher, Event::FileReadyCb cb,
+                                               uint32_t events, UserspaceIoHandle& io_source)
+    : schedulable_(dispatcher.createSchedulableCallback([this, cb]() {
+        auto ephemeral_events = event_listener_.getAndClearEphemeralEvents();
+        ENVOY_LOG(trace, "User space event {} invokes callbacks on events = {}",
+                  static_cast<void*>(this), ephemeral_events);
+        cb(ephemeral_events);
+      })),
+      io_source_(io_source) {
+  setEnabled(events);
+}
+
+void EventListenerImpl::clearEphemeralEvents() {
+  // Clear ephemeral events to align with FileEventImpl::setEnable().
+  ephemeral_events_ = 0;
+}
+
+void EventListenerImpl::onEventActivated(uint32_t activated_events) {
+  ephemeral_events_ |= activated_events;
+}
+
+void EventListenerImpl::setEnabledEvents(uint32_t enabled_events) {
+  enabled_events_ = enabled_events;
+}
+
+void UserSpaceFileEventImpl::activate(uint32_t events) {
+  // Only supported event types are set.
+  ASSERT((events & (Event::FileReadyType::Read | Event::FileReadyType::Write |
+                    Event::FileReadyType::Closed)) == events);
+  event_listener_.onEventActivated(events);
+  schedulable_->scheduleCallbackNextIteration();
+}
+
+void UserSpaceFileEventImpl::setEnabled(uint32_t events) {
+  // Only supported event types are set.
+  ASSERT((events & (Event::FileReadyType::Read | Event::FileReadyType::Write |
+                    Event::FileReadyType::Closed)) == events);
+  event_listener_.clearEphemeralEvents();
+  event_listener_.setEnabledEvents(events);
+  bool was_enabled = schedulable_->enabled();
+  // Recalculate activated events.
+  uint32_t events_to_notify = 0;
+  if ((events & Event::FileReadyType::Read) && io_source_.isReadable()) {
+    events_to_notify |= Event::FileReadyType::Read;
+  }
+  if ((events & Event::FileReadyType::Write) && io_source_.isPeerWritable()) {
+    events_to_notify |= Event::FileReadyType::Write;
+  }
+  if ((events & Event::FileReadyType::Closed) && io_source_.isPeerShutDownWrite()) {
+    events_to_notify |= Event::FileReadyType::Closed;
+  }
+  if (events_to_notify != 0) {
+    activate(events_to_notify);
+  } else {
+    schedulable_->cancel();
+  }
+  ENVOY_LOG(
+      trace,
+      "User space file event {} set enabled events {} and events {} is active. Will {} reschedule.",
+      static_cast<void*>(this), events, was_enabled ? "not " : "");
+}
+
+void UserSpaceFileEventImpl::poll(uint32_t events) {
+  ASSERT((events & (Event::FileReadyType::Read | Event::FileReadyType::Write |
+                    Event::FileReadyType::Closed)) == events);
```

Comment:
> Event::FileReadyType::Closed is never passed in to this method.

setEnabled should check if `Closed` has occurred.

Reply:
> setEnabled should check if `Closed` has occurred.

I was wrong about ^^.

setEnabled() could use this. But it doesn't and avoid checking enabled events twice.

Reply:
Anyway, setNewDataAvailable() should activate `Closed` to emulate shutdown(SHUT_WR) at peer so activateIfEnabled() must accept `Closed`. WDYT?

Reply:
Also, since activateIfEnabled() accept `Closed`, the activate() invoked by `activateIfEnabled()` should also accept `Closed`. Although `Network::Connection` may not directly call `activate(Closed)` as you described

Reply:
The delivery of Closed events doesn't work on real sockets in some platforms.  I think use of Closed should be removed across the board eventually, but I guess for now it seems sensible to accept Closed for activateIfEnabled, but not for activate.  Seems like #13418 handles shutdown correctly by doing activateIfEnabled(Read | Closed); Closed is not always registered, sometimes you need to rely on the read registration to be the one that carries the event across.

Reply:
> The delivery of Closed events doesn't work on real sockets in some platforms.
That's why the initial commits do not handle `Closed` :)

Ideally, the event owner (e.g. connection) can ask for `Closed` and consider `Closed` event as optimization. Especially when the L7 protocol doesn't have to drain the io_handle buffer. We can put this long term goal aside.

Even the connection goes as you described, this extra `Closed` delivery is not breaking the connection, since there are some platforms (mostly importantly, linux) could deliver `Closed` as this userspace file event :)